kv/etcd/etcd.go

```diff
@@ -0,0 +1,220 @@
+	"sync"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/protobuf/proto"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"golang.org/x/net/context"
+)
+
+var noopCancel func()
+
+// NewStore creates a kv store based on etcd
+func NewStore(opts Options) (kv.Store, error) {
+	etcd, err := clientv3.New(clientv3.Config{
+		Endpoints:   opts.Endpoints(),
+		DialTimeout: opts.DialTimeout(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		opts:       opts,
+		kv:         etcd.KV,
+		watcher:    etcd.Watcher,
+		watchables: map[string]kv.ValueWatchable{},
+		retrier:    xretry.NewRetrier(opts.RetryOptions()),
+		logger:     opts.InstrumentsOptions().Logger(),
+	}, nil
+}
+
+type client struct {
+	sync.RWMutex
+
+	opts       Options
+	kv         clientv3.KV
+	watcher    clientv3.Watcher
+	watchables map[string]kv.ValueWatchable
+	retrier    xretry.Retrier
+	logger     xlog.Logger
+}
+
+func (c *client) Get(key string) (kv.Value, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	r, err := c.kv.Get(ctx, c.opts.KeyFn()(key))
+	if err != nil {
+		return nil, err
+	}
+
+	if r.Count == 0 {
+		return nil, kv.ErrNotFound
+	}
+
+	if r.Count > 1 {
+		return nil, fmt.Errorf("received %d values for key %s, expecting 1", r.Count, key)
+	}
+
+	return kv.NewValue(r.Kvs[0].Value, int(r.Kvs[0].Version)), nil
+}
+
+func (c *client) Watch(key string) (kv.ValueWatch, error) {
+	c.Lock()
+	defer c.Unlock()
+	watchable, ok := c.watchables[key]
+	if !ok {
+		watchChan := c.watcher.Watch(
+			context.Background(),
+			c.opts.KeyFn()(key),
+			clientv3.WithProgressNotify(),
+			// Receive initial notification once the watch channel is created
+			clientv3.WithCreatedNotify(),
+		)
+
+		watchable = kv.NewValueWatchable()
+		c.watchables[key] = watchable
+
+		go func() {
+			for r := range watchChan {
+				err := r.Err()
+				if err != nil {
+					c.logger.Errorf("received error on watch channel: %v", err)
+				}
+
+				// we need retry here because if Get() failed on an watch update,
+				// it has to wait 10 mins to be notified to try again
+				err = c.retrier.Attempt(func() error {
+					return c.update(watchable, key)
+				})
+
+				if err != nil {
+					c.logger.Errorf("received notification for key %s, but failed to get value: %v", key, err)
+				}
+			}
+			c.logger.Warnf("no longer watching key: %s", key)
+		}()
+	}
+
+	_, w, err := watchable.Watch()
+	return w, err
+}
+
+func (c *client) update(w kv.ValueWatchable, key string) error {
+	v, err := c.Get(key)
+	if err != nil {
+		return err
+	}
+
+	curValue := w.Get()
+	if curValue != nil && curValue.Version() >= v.Version() {
+		return nil
+	}
+
+	return w.Update(v)
+}
+
+func (c *client) Set(key string, v proto.Message) (int, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	value, err := proto.Marshal(v)
+	if err != nil {
+		return 0, err
+	}
+
+	r, err := c.kv.Put(ctx, c.opts.KeyFn()(key), string(value), clientv3.WithPrevKV())
```

Comment:
Hm might not be great to store the raw bytes as `string` here, `etcd` has an HTTP API and it may not play nicely with this API to use raw bytes as string.  Probably more prudent to base64 the bytes into a string, then decoded base64 before unmarshalling when retrieving a value.